File: ts/test/cssparsing.test.ts
/* Copyright (c) 2026 Richard Rodger and other contributors, MIT License */

// Conformance against the THIRD-PARTY CSS Syntax Level 3 corpus
// (SimonSapin/css-parsing-tests -- the implementation-independent suite used
// by rust-cssparser/Servo, tinycss2 and Crass).
//
// The corpus is NOT vendored. `scripts/fetch-css-parsing-tests.sh` clones it
// at a pinned commit into `test/css-parsing-tests/`, which is gitignored. The
// `pretest` npm script runs it, so a normal `npm test` always has it. If it is
// absent this file FAILS LOUDLY -- it never skips.
//
// `go/cssparsing_test.go` runs the SAME files with the SAME derivation.
//
// ---------------------------------------------------------------------------
// WHAT IS COMPARED, AND WHY
// ---------------------------------------------------------------------------
// Upstream's expected values are component-value trees: a token-level model
// (idents, delims, whitespace, nested {} [] () blocks) that this plugin does
// not produce and does not claim to. It produces a reworkcss-style rule AST.
// So the runner compares a *kind sequence* computed MECHANICALLY from
// upstream's own expected output:
//
//   ["at-rule", name, ...]      -> "@" + name
//   ["qualified rule", ...]     -> "rule"
//   ["declaration", name, ...]  -> "decl:" + name
//   ["error", ...]              -> "error"
//
// and the same sequence read off this plugin's AST (comment nodes dropped --
// CSS Syntax discards comments, so upstream's model has no item to match).
//
// If upstream's expected sequence contains an "error" item, the input
// contains a construct that CSS Syntax reports as a parse error. This plugin
// has no error-recovery mode (unlike reworkcss's `silent` option), so the only
// conformant response available to it is to REJECT the whole input. Those
// cases are therefore asserted as must-reject.
//
// FILES DELIBERATELY NOT WIRED, and why -- these are NOT a skip list, they
// test algorithms this plugin has no entry point for, so running them would
// manufacture failures that say nothing about CSS conformance:
//   one_rule.json / one_declaration.json  "parse a rule"/"parse a declaration"
//       parse exactly ONE item and report "extra-input" for anything after it;
//       a stylesheet parser legitimately accepts several. Their `;` semantics
//       also differ from declaration-list semantics.
//   component_value_list.json / one_component_value.json
//       token-level, no rule/declaration structure to compare.
//   stylesheet_bytes.json   byte-stream encoding detection; this plugin takes
//       a decoded string.
//   color_*.json / An+B.json
//       value-level micro-syntaxes; this plugin keeps declaration values as
//       raw text by design (documented in README.md).
//
// There is no SKIP list among the wired files: every case in every wired file
// runs and is asserted.

import { describe, test } from 'node:test'
import assert from 'node:assert'
import { existsSync, readFileSync } from 'node:fs'
import { join } from 'node:path'

import { Tabnas } from '@tabnas/parser'
import { jsonic } from '@tabnas/jsonic'
import { Css } from '../dist/css'

// Pinned upstream. Keep in step with scripts/fetch-css-parsing-tests.sh.
const UPSTREAM = 'https://github.com/SimonSapin/css-parsing-tests'
const SHA = '203ce36bffd617db7f118c551e32794561fb273d'

const suiteDir = join(__dirname, '..', '..', 'test', 'css-parsing-tests')

function requireCorpus(): void {
  if (existsSync(join(suiteDir, 'stylesheet.json'))) return
  throw new Error(
    'MISSING CONFORMANCE CORPUS: ' +
      suiteDir +
      '\nThe css-parsing-tests corpus is third-party and is never committed.' +
      '\nFetch it (pinned to ' +
      SHA +
      ') with:\n\n    bash scripts/fetch-css-parsing-tests.sh\n\n' +
      'This test does NOT skip when the corpus is absent: a conformance ' +
      'suite that quietly does not run is worse than no suite at all.',
  )
}

function parseCss(src: string): any {
  return new Tabnas().use(jsonic).use(Css).parse(src)
}

// Each upstream file is a flat array of [input, expected, input, expected...].
function loadPairs(file: string): { input: string; expected: any[] }[] {
  const flat = JSON.parse(readFileSync(join(suiteDir, file), 'utf8'))
  const out: { input: string; expected: any[] }[] = []
  for (let i = 0; i + 1 < flat.length; i += 2) {
    out.push({ input: flat[i], expected: flat[i + 1] })
  }
  return out
}

function upstreamKind(item: any): string {
  if (!Array.isArray(item)) return 'other:' + JSON.stringify(item)
  switch (item[0]) {
    case 'at-rule':
      return '@' + item[1]
    case 'qualified rule':
      return 'rule'
    case 'declaration':
      return 'decl:' + item[1]
    case 'error':
      return 'error'
    default:
      return 'other:' + JSON.stringify(item[0])
  }
}

function nodeKind(node: any): string | null {
  const t = node && node.type
  if ('comment' === t) return null
  if ('rule' === t) return 'rule'
  if ('declaration' === t) return 'decl:' + node.property
  // Every other node type in this AST is an at-rule; `vendor` is split off
  // the keyword by the grammar, so put it back to recover the at-keyword.
  return '@' + (node.vendor ?? '') + t
}

function nodeKinds(nodes: any[]): string[] {
  return (nodes ?? []).map(nodeKind).filter((k): k is string => null !== k)
}

function label(s: string): string {
  const one = JSON.stringify(s)
  return 60 < one.length ? one.slice(0, 57) + '..."' : one
}

// `wrap` turns a declaration-list/block-contents input into a whole
// stylesheet; `pick` pulls the matching node array back out of the result.
function runFile(
  file: string,
  count: number,
  wrap: (s: string) => string,
  pick: (ast: any) => any[],
) {
  requireCorpus()
  const pairs = loadPairs(file)

  describe(file, () => {
    test('corpus census', () => {
      assert.equal(
        pairs.length,
        count,
        `expected ${count} cases in ${file} at ${UPSTREAM}@${SHA}; the ` +
          'pinned corpus changed. Do not silently accept a different suite.',
      )
    })

    for (const p of pairs) {
      const wantKinds = p.expected.map(upstreamKind)
      const mustReject = wantKinds.includes('error')

      test(`${mustReject ? '[reject]' : '[accept]'} ${label(p.input)}`, () => {
        const src = wrap(p.input)

        if (mustReject) {
          assert.throws(
            () => parseCss(src),
            'CSS Syntax L3 reports a parse error for this input ' +
              `(upstream expects ${JSON.stringify(wantKinds)}); this plugin ` +
              'has no error-recovery mode, so it must reject.',
          )
          return
        }

        const raw = parseCss(src)
        // An empty stylesheet parses to undefined by design (documented in
        // AGENTS.md), and upstream expects [] for it -- that is a match.
        const got = undefined === raw ? { rules: [] } : JSON.parse(JSON.stringify(raw))
        assert.deepStrictEqual(nodeKinds(pick(got)), wantKinds, label(p.input))
      })
    }
  })
}

const identity = (s: string) => s
const topRules = (ast: any) => ast.rules ?? []
// A declaration list / block contents has no standalone entry point here, so
// wrap it in a style rule and read that rule's `declarations` back out. CSS
// Nesting means nested rules and at-rules also land in `declarations`, which
// is exactly what blocks_contents.json expects.
const wrapBlock = (s: string) => 'a{\n' + s + '\n}'
const blockDecls = (ast: any) => ast.rules?.[0]?.declarations ?? []

describe(`css-parsing-tests (CSS Syntax L3, ${SHA.slice(0, 7)})`, () => {
  test('corpus is present', () => requireCorpus())

  runFile('stylesheet.json', 16, identity, topRules)
  runFile('rule_list.json', 15, identity, topRules)
  runFile('declaration_list.json', 10, wrapBlock, blockDecls)
  runFile('blocks_contents.json', 13, wrapBlock, blockDecls)
})

File: go/cssparsing_test.go
// Copyright (c) 2026 Richard Rodger and other contributors, MIT License

package tabnascss

// cssparsing_test.go — conformance against the THIRD-PARTY CSS Syntax Level 3
// corpus (SimonSapin/css-parsing-tests: the implementation-independent suite
// used by rust-cssparser/Servo, tinycss2 and Crass).
//
// Mirrors ts/test/cssparsing.test.ts exactly: same files, same derivation,
// same census, same must-reject verdict. Read the long comment at the top of
// that file (or test/AGENTS.md) for WHY a kind sequence is compared and which
// upstream files are deliberately not wired — those exclusions are reasoned,
// not a skip list. Among the wired files every case runs and is asserted.
//
// The corpus is never committed; TestMain fetches it at a pinned commit. If
// it is absent these tests FAIL LOUDLY — they never skip.

import (
	"encoding/json"
	"os"
	"path/filepath"
	"reflect"
	"strconv"
	"testing"
)

// Pinned upstream. Keep in step with scripts/fetch-css-parsing-tests.sh and
// ts/test/cssparsing.test.ts.
const cpUpstream = "https://github.com/SimonSapin/css-parsing-tests"
const cpSHA = "203ce36bffd617db7f118c551e32794561fb273d"

func cpDir() string { return filepath.Join(repoRoot(), "test", "css-parsing-tests") }

func cpRequire(t *testing.T) {
	t.Helper()
	if _, err := os.Stat(filepath.Join(cpDir(), "stylesheet.json")); err != nil {
		t.Fatal(corpusMissing(cpDir(), cpSHA, "scripts/fetch-css-parsing-tests.sh"))
	}
}

type cpPair struct {
	input    string
	expected []any
}

func cpLoad(t *testing.T, file string) []cpPair {
	t.Helper()
	raw, err := os.ReadFile(filepath.Join(cpDir(), file))
	if err != nil {
		t.Fatalf("read %s: %v", file, err)
	}
	var flat []any
	if err := json.Unmarshal(raw, &flat); err != nil {
		t.Fatalf("parse %s: %v", file, err)
	}
	var out []cpPair
	for i := 0; i+1 < len(flat); i += 2 {
		in, ok := flat[i].(string)
		if !ok {
			t.Fatalf("%s: pair %d input is not a string", file, i/2)
		}
		exp, ok := flat[i+1].([]any)
		if !ok {
			t.Fatalf("%s: pair %d expected is not an array", file, i/2)
		}
		out = append(out, cpPair{input: in, expected: exp})
	}
	return out
}

func cpUpstreamKind(item any) string {
	arr, ok := item.([]any)
	if !ok || len(arr) == 0 {
		return "other:" + mustJSON(item)
	}
	head, _ := arr[0].(string)
	switch head {
	case "at-rule":
		name, _ := arr[1].(string)
		return "@" + name
	case "qualified rule":
		return "rule"
	case "declaration":
		name, _ := arr[1].(string)
		return "decl:" + name
	case "error":
		return "error"
	}
	return "other:" + mustJSON(head)
}

// cpNodeKind maps one tabnas AST node to a kind. Comment nodes return "" and
// are dropped: CSS Syntax discards comments, so upstream has no item to match.
func cpNodeKind(node any) string {
	m, ok := node.(map[string]any)
	if !ok {
		return "other:" + mustJSON(node)
	}
	typ, _ := m["type"].(string)
	switch typ {
	case "comment":
		return ""
	case "rule":
		return "rule"
	case "declaration":
		prop, _ := m["property"].(string)
		return "decl:" + prop
	}
	vendor, _ := m["vendor"].(string)
	return "@" + vendor + typ
}

func cpNodeKinds(nodes any) []string {
	out := []string{}
	arr, ok := nodes.([]any)
	if !ok {
		return out
	}
	for _, n := range arr {
		if k := cpNodeKind(n); k != "" {
			out = append(out, k)
		}
	}
	return out
}

func cpLabel(s string) string {
	q := strconv.Quote(s)
	if 60 < len(q) {
		return q[:57] + `..."`
	}
	return q
}

// cpRunFile: `wrap` turns a declaration-list/block-contents input into a whole
// stylesheet; `pick` pulls the matching node array back out of the result.
func cpRunFile(
	t *testing.T,
	file string,
	count int,
	wrap func(string) string,
	pick func(map[string]any) any,
) {
	t.Helper()
	cpRequire(t)
	pairs := cpLoad(t, file)

	t.Run(file+"/corpus census", func(t *testing.T) {
		if len(pairs) != count {
			t.Fatalf("expected %d cases in %s at %s@%s, found %d; the pinned "+
				"corpus changed. Do not silently accept a different suite.",
				count, file, cpUpstream, cpSHA, len(pairs))
		}
	})

	for _, p := range pairs {
		p := p
		want := []string{}
		mustReject := false
		for _, item := range p.expected {
			k := cpUpstreamKind(item)
			if k == "error" {
				mustReject = true
			}
			want = append(want, k)
		}

		verdict := "[accept] "
		if mustReject {
			verdict = "[reject] "
		}
		t.Run(file+"/"+verdict+cpLabel(p.input), func(t *testing.T) {
			src := wrap(p.input)
			got, err := parseCss(src, map[string]any{})

			if mustReject {
				if err == nil {
					t.Fatalf("CSS Syntax L3 reports a parse error for this "+
						"input (upstream expects %v); this plugin has no "+
						"error-recovery mode, so it must reject. Got: %s",
						want, mustJSON(got))
				}
				return
			}

			if err != nil {
				t.Fatalf("unexpected parse error: %v", err)
			}
			// An empty stylesheet parses to nil by design (AGENTS.md), and
			// upstream expects [] for it — that is a match.
			nodes := any([]any{})
			if got != nil {
				norm, nerr := jsonNorm(got)
				if nerr != nil {
					t.Fatalf("%v", nerr)
				}
				m, ok := norm.(map[string]any)
				if !ok {
					t.Fatalf("parse produced %#v, not a stylesheet", norm)
				}
				nodes = pick(m)
			}
			if g := cpNodeKinds(nodes); !reflect.DeepEqual(g, want) {
				t.Errorf("%s:\n  got  %v\n  want %v", cpLabel(p.input), g, want)
			}
		})
	}
}

func cpIdentity(s string) string  { return s }
func cpWrapBlock(s string) string { return "a{\n" + s + "\n}" }

func cpTopRules(ast map[string]any) any { return ast["rules"] }

// A declaration list / block contents has no standalone entry point here, so
// it is wrapped in a style rule and that rule's `declarations` read back out.
// CSS Nesting means nested rules and at-rules also land in `declarations`,
// which is exactly what blocks_contents.json expects.
func cpBlockDecls(ast map[string]any) any {
	rules, ok := ast["rules"].([]any)
	if !ok || len(rules) == 0 {
		return []any{}
	}
	first, ok := rules[0].(map[string]any)
	if !ok {
		return []any{}
	}
	return first["declarations"]
}

func TestCssParsingTests(t *testing.T) {
	cpRequire(t)
	cpRunFile(t, "stylesheet.json", 16, cpIdentity, cpTopRules)
	cpRunFile(t, "rule_list.json", 15, cpIdentity, cpTopRules)
	cpRunFile(t, "declaration_list.json", 10, cpWrapBlock, cpBlockDecls)
	cpRunFile(t, "blocks_contents.json", 13, cpWrapBlock, cpBlockDecls)
}
